From f0d161cfa9d79a32c851288773fae7a53ef5bb87 Mon Sep 17 00:00:00 2001
From: Naveen S <quic_naves@quicinc.com>
Date: Tue, 30 Jan 2024 10:31:03 +0530
Subject: [PATCH] wifi: mac80211: Set eht operation tag's basic mcs nss value.

Currently the ext tag: eht oper's basic mcs nss value is set to the
max tx and rx nss values. This is an issue because if the station's
rx and tx nss are below the basic mcs nss of the eht oper in AP, the
association will not happend in EHT mode(as per draft doc at EOM).

Hence the basic mcs nss is now set to be 0x00000011 instead of the max
nss values, so that the association between station and AP can now
be done with the EHT mode.

w.r.t. the draft doc.: IEEE 802.11-23/0638r4 "An EHT STA shall not
attempt to join (MLME-JOIN.request primitive) a BSS unless it supports
(i.e., is able to both transmit and receive using) all of the
<EHT-MCS, NSS> tuples in the basic EHT-MCS and NSS set."

Signed-off-by: Naveen S <quic_naves@quicinc.com>
---
 include/linux/ieee80211.h | 3 +++
 net/mac80211/util.c       | 4 +---
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/include/linux/ieee80211.h b/include/linux/ieee80211.h
index dd21225..abfe832 100644
--- a/include/linux/ieee80211.h
+++ b/include/linux/ieee80211.h
@@ -2139,6 +2139,9 @@ struct ieee80211_eht_cap_elem {
 #define IEEE80211_EHT_OPERATION_FIXED_LEN			0x5
 #define IEEE80211_EHT_OPERATION_INFO_FIXED_LEN			0x3
 #define DISABLED_SUBCHANNEL_BITMAP_BYTES_SIZE			0x2
+
+#define EHT_BASIC_MCS_NSS_SET					0x11
+
 /**
  * struct ieee80211_eht_operation - eht operation element
  *
diff --git a/net/mac80211/util.c b/net/mac80211/util.c
index ea5ca87..3ac8f2f 100644
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -3703,8 +3703,6 @@ u8 *ieee80211_ie_build_eht_oper(u8 *pos, struct cfg80211_chan_def *chandef,
 				const struct ieee80211_sta_eht_cap *eht_cap)
 
 {
-	const struct ieee80211_eht_mcs_nss_supp_20mhz_only *eht_mcs_nss =
-					&eht_cap->eht_mcs_nss_supp.only_20mhz;
 	struct ieee80211_eht_operation *eht_oper;
 	struct ieee80211_eht_operation_info *eht_oper_info;
 	u8 eht_oper_len = IEEE80211_EHT_OPERATION_FIXED_LEN;
@@ -3722,7 +3720,7 @@ u8 *ieee80211_ie_build_eht_oper(u8 *pos, struct cfg80211_chan_def *chandef,
 
 	eht_oper = (struct ieee80211_eht_operation *)pos;
 
-	memcpy(&eht_oper->basic_mcs_nss, eht_mcs_nss, sizeof(*eht_mcs_nss));
+	eht_oper->basic_mcs_nss.rx_tx_mcs7_max_nss = EHT_BASIC_MCS_NSS_SET;
 	eht_oper->params |= IEEE80211_EHT_OPER_INFO_PRESENT;
 	pos += eht_oper_len;
 
-- 
2.17.1

